packages/mui-material/src/utils/useSlot.ts

```diff
@@ -0,0 +1,172 @@
+'use client';
+import * as React from 'react';
+import { ClassValue } from 'clsx';
+import { unstable_useForkRef as useForkRef } from '@mui/utils';
+import { appendOwnerState, resolveComponentProps, mergeSlotProps } from '@mui/base/utils';
+
+export type WithCommonProps<T> = T & {
+  className?: string;
+  style?: React.CSSProperties;
+  ref?: React.Ref<any>;
+};
+
+type EventHandlers = Record<string, React.EventHandler<any>>;
+
+type ExtractComponentProps<P> = P extends infer T | ((ownerState: any) => infer T) ? T : never;
+
+/**
+ * An internal function to create a Material UI slot.
+ *
+ * This is an advanced version of Base UI `useSlotProps` because Material UI allows leaf component to be customized via `component` prop
```

Comment:
I don't know whether we should keep the lead component functionality. What do you think? It seems to me that it doesn't harm, but I'm unsure if it could have unwanted secondary effects.